### PR TITLE
Export AjaxException members to better support Future.toPromise and Ajax Futures.

### PR DESCRIPTION
**Pull request by russwyte**
When using the rich future JSConverter to return a JS Promise from a Future that performs an AJAX request it is possible that the JS caller will be returned an AjaxException to handle in a rejected promise. In these cases is it useful for the JS client to have access to the XHR request itself to better handle the failure. This change simply makes the XHR that caused the exception available to the JS caller.

### CODE REVIEW

**Comment by sjrd** [comment]
What's the reason for `@JSExport`? I understand from the PR message (which should be in the commit message, btw) that `@JSExportAll` is useful, but not `@JSExport`.

Exports in libraries should be used with great care, as they mean code that will always be included in the emitted JS files of any application using that library, even if they are not used.

**Comment by russwyte** [comment]
Thanks. 

Good points. I will add that kind of detail to future PR commits. 

My mistake on @JSExport I thought it was always required when exporting members.

File: src/main/scala/org/scalajs/dom/ext/Extensions.scala [hunk 1]
```diff
@@ -3,16 +3,16 @@ package org.scalajs.dom.ext
 import java.nio.ByteBuffer
 
 import scala.language.implicitConversions
-import scala.concurrent.{Promise, Future}
-
+import scala.concurrent.{Future, Promise}
 import scala.scalajs.js
 import scala.scalajs.js.typedarray._
 import scala.scalajs.js.typedarray.TypedArrayBufferOps._
-
```
Comment (sjrd) [inline src/main/scala/org/scalajs/dom/ext/Extensions.scala]: The changes here and just above are useless and should not be part of this commit/PR.

File: src/main/scala/org/scalajs/dom/ext/Extensions.scala [hunk 2]
```diff
@@ -234,6 +234,7 @@ object KeyCode {
  * Thrown when `Ajax.get` or `Ajax.post` receives a non-20X response code.
  * Contains the XMLHttpRequest that resulted in that response
  */
+@JSExportAll
 case class AjaxException(xhr: dom.XMLHttpRequest) extends Exception {
   def isTimeout = xhr.status == 0 && xhr.readyState == 4
```
Comment (sjrd) [inline src/main/scala/org/scalajs/dom/ext/Extensions.scala]: In fact, I now realize this a `case class`. Case classes have several "hidden" public members such as `copy`, `hashCode`, which we do not really want to be exported. I recommend to only export `xhr` itself. Possibly `isTimeout`, but I would refrain from doing so, in order to minimize what is exported.

**Comment by sjrd** [comment]
I'm wondering how much this is really needed/appropriate for this library. scalajs-dom is not really targeted at providing a nice *JavaScript* API for its things. I'm worried that including this change might set a precedent for exporting stuff, which would then justify exporting more stuff in all kinds of place in this API. Eventually leading to code size bloat for users of the library that do not even care about exporting a JS API.

IMO this kind of features should be left to applications or dedicated libraries. For this particular use case, I would recommend something like the following user-space pimp:
```scala
implicit class AjaxExceptionFutureOps(val self: Future[AjaxException]) extends AnyVal {
  def toXHRPromise: js.Promise[XMLHttpRequest] =
    self.transform(identity, ex => js.JavaScriptException(ex.xhr))
}
```
You can then use `fut.toXHRPromise` to get a `js.Promise[XMLHttpRequest]`, without needing any export. This is also slightly more idiomatic, IMO: it is often better to provide *conversions* from Scala types to JS types, rather than exporting JS APIs in Scala types. Except of course for Scala classes that are *dedicated* to provide a JS API, but that is not the case of `AjaxException`.

**Comment by russwyte** [comment]
I actually like your last suggestion very much. It might be preferable if the default toPromise transformation worked like that - because the AjaxException does not have a lot of value as exported JS - and that is the whole point of toPromise.

**Comment by sjrd** [comment]
> It might be preferable if the default toPromise transformation worked like that

Hum ... no, it wouldn't. `.toJSPromise` converts the `Future` into a `js.Promise`. It does not touch the wrapped value. That's the same as all other conversions in the world, e.g., `seq.toJSArray`. Even if it were desirable, you cannot have `.toJSPromise` deal with all the possible "convenient" deep conversions in the world and beyond.

**Comment by russwyte** [comment]
I only meant when used with a future obtained via Ajax.get .post etc... Not for all Futures.

**Comment by sjrd** [comment]
How would you implement that?

**Comment by russwyte** [comment]
I might change to this impl of "onreadystatechange":

`    req.onreadystatechange = {(e: dom.Event) =>
      if (req.readyState.toInt == 4){
        if ((req.status >= 200 && req.status < 300) || req.status == 304)
          promise.success(req)
        else
          promise.failure(js.JavaScriptException(req))
      }
    }`

Because we know a js.Promise is something only external JS cares about we should not use type that is obtuse to JS like the current AjaxException - it carries no useful information about the failure - but instead something with the XHR accessible.

**Comment by sjrd** [comment]
But most people use `Ajax.get` from *Scala* code and want to see a Scala exception that can be pattern matched against in a `catch` clause. Your use case is comparatively much rarer.

**Comment by russwyte** [comment]
Yes - I actually agree with you. I think for now I will go the implicit conversion route you described earlier - but I think it would be great to expose a bit more (at least the xhr) to the JS clients without hoops if possible. If I can think of a better way to do it I will propose it later.

Thanks for the information and insight into the library strategy.
